lib/semantics/check-omp-clause.h

```diff
@@ -0,0 +1,111 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// OpenMP clause validity check for directives
+
+#ifndef FORTRAN_SEMANTICS_CHECK_OMP_CLAUSE_H_
+#define FORTRAN_SEMANTICS_CHECK_OMP_CLAUSE_H_
+
+#include "semantics.h"
+#include "../parser/parse-tree.h"
+#include <typeindex>
+
+namespace Fortran::semantics {
+
+class OmpClauseChecker : public virtual BaseChecker {
+public:
+  OmpClauseChecker(SemanticsContext &context) : context_{context} {}
+
+  void Enter(const parser::OmpBlockDirective &);
+  void Enter(const parser::OmpBlockDirective::Parallel &);
+  void Leave(const parser::OmpEndBlockDirective &);
+
+  void Enter(const parser::OmpClause &);
+  void Enter(const parser::OmpClause::Defaultmap &);
+  void Enter(const parser::OmpClause::Inbranch &);
+  void Enter(const parser::OmpClause::Mergeable &);
+  void Enter(const parser::OmpClause::Nogroup &);
+  void Enter(const parser::OmpClause::Notinbranch &);
+  void Enter(const parser::OmpClause::Untied &);
+  void Enter(const parser::OmpClause::Collapse &);
+  void Enter(const parser::OmpClause::Copyin &);
+  void Enter(const parser::OmpClause::Copyprivate &);
+  void Enter(const parser::OmpClause::Device &);
+  void Enter(const parser::OmpClause::DistSchedule &);
+  void Enter(const parser::OmpClause::Final &);
+  void Enter(const parser::OmpClause::Firstprivate &);
+  void Enter(const parser::OmpClause::From &);
+  void Enter(const parser::OmpClause::Grainsize &);
+  void Enter(const parser::OmpClause::Lastprivate &);
+  void Enter(const parser::OmpClause::NumTasks &);
+  void Enter(const parser::OmpClause::NumTeams &);
+  void Enter(const parser::OmpClause::NumThreads &);
+  void Enter(const parser::OmpClause::Ordered &);
+  void Enter(const parser::OmpClause::Priority &);
+  void Enter(const parser::OmpClause::Private &);
+  void Enter(const parser::OmpClause::Safelen &);
+  void Enter(const parser::OmpClause::Shared &);
+  void Enter(const parser::OmpClause::Simdlen &);
+  void Enter(const parser::OmpClause::ThreadLimit &);
+  void Enter(const parser::OmpClause::To &);
+  void Enter(const parser::OmpClause::Uniform &);
+  void Enter(const parser::OmpClause::UseDevicePtr &);
+
+  void Enter(const parser::OmpAlignedClause &);
+  void Enter(const parser::OmpDefaultClause &);
+  void Enter(const parser::OmpDependClause &);
+  void Enter(const parser::OmpIfClause &);
+  void Enter(const parser::OmpLinearClause &);
+  void Enter(const parser::OmpMapClause &);
+  void Enter(const parser::OmpProcBindClause &);
+  void Enter(const parser::OmpReductionClause &);
+  void Enter(const parser::OmpScheduleClause &);
+
```

Comment:
Everything from here down should be private. And given that, I don't think there is any benefits to having these member functions to set the data members: just assign to the data members directly. This doesn't apply to `CheckAllowed` because it does more than a simple assignment.